pyproject.toml

```diff
@@ -52,13 +53,12 @@ prefixcommons = ">=0.1.12"
 pyyaml = "*"
 rdflib = ">=6.0.0"
 requests = "*"
-prefixmaps = ">=0.1.4"
+prefixmaps = ">=0.2.2"
```

Comment:
We have the version of `prefixmaps` specified at `0.2.2` in `poetry.lock`. That's good! Our local dev environments and CI environments will be able to test under Python 3.8 through Python 3.12. But I think we _shouldn't_ change this specification in `pyproject.toml`. This change would mean all clients of `linkml` need to use a very recent version of `prefixmaps`, causing potential dependency incompatibilities. A client of `linkml` should be able to continue using `prefixmaps < 0.2.2` in their environment.

Reply:
Wasnt that update to fix some incompatibility between 3.8 and 3.12? So the dependency should be like >=0.1.4 for python<3.12 and >=0.2.2 for 3.12?

Reply:
To avoid the long feedback round-trips due to time-shift, I'm applying my proposal on the branch with the hope that it's addressing the concerns of this conversation. I'm not merging though before getting the confirmation from @pkalita-lbl

Reply:
Sorry I'm not trying to drag this out, but this is still over-constraining. If a client is using Python 3.12 there's no reason they need to also have `prefixmaps >= 0.2.2`. Versions of `prefixmaps` going back to 0.2.0 declare support for Python 3.12. All `linkml` needs to do is say "we need `prefixmaps >= 0.1.4`" as we did before. If a client is using Python 3.12, it's up to their package installation tool (whether that's pip or Poetry or whatever) to resolve an appropriate version of `prefixmaps` for their environment.

When it comes to _testing_, we will get a version (0.2.2) that's compatible with all the versions of Python we want to test with because of what's specified in `poetry.lock`.

Reply:
Now I got your point.

Even if `linkml-runtime` declares that `prefixmaps` should be "only" `>=0.1.4` on Python 3.12, the `prefixmaps` package itself already does so. It means that, the moment that Pip or Poetry tries to resolve dependencies, it realizes that it needs `prefixmaps` `>=0.2.2`. You're fully right!

I'm dropping my last commit and removing the change in the `pyproject.toml` WRT `prefixmaps`.

Reply:
Yes you got it! I'll keep an eye out for that commit so we can finally merge this work!

Reply:
It should be done now. Please have a look at it and feel free to either merge yourself or approve (then I'd merge).

@pkalita-lbl thanks for your valuable input!

Reply:
i'm also not trying to be be difficult, and ultimately think >=0.1.4 would be sufficient in any first-order dependency situation by virtue of how `pip` and `poetry` will both resolve to the most recent version available and not try and install `<0.2`, but since `prefixmaps<2` declare python compatibility `^3.7.6` it is still possible for a 3.12 project to resolve to `prefixmaps<2`. This would be a problem with a package that depends on `linkml` and any other project that specifies `<0.2` (or `^0.1.*`, etc.). Thinking about bioregistry and pydantic 1/2 problems here. 

arguably the ideal thing would be for `prefixmaps<0.2` to specify `python<3.12` if it's known to not work, but unless we want to issue a bunch of `postN` releases (which i hope we don't) it doesn't. but if we know that `<0.2` doesn't work for `3.12` then we should spec that here - dependency resolution should correctly fail for a project that tries to use `prefixmaps<0.2` in 3.12. 

The role of version specifiers is to exclude combinations of dependencies known to be incompatible through an entire dependency tree - we could be maximally unrestrictive and let people YOLO their own environments entirely and leave all version specifiers blank, but then suffer the consequences of everyone's environment only working by chance. IMO it's good practice to exclude known bad dep combos, in this case we're not excluding any valid environments, but preventing unforseen footguns.

ultimately not my call! just looking out for future issue triagers <3

Reply:
oh sorry i didn't refresh my page and see that you had responded and also resolved this thread. my bad! whatevs, i'm still lgtm and maybe this is a more general discussion about how to manage deps

Reply:
I totally understand what you're saying.

Just stepping back, in the past I think `linkml` has erred (intentionally or not) on the side of being too restrictive in its dependency specifications. And it has led to downstream clients reporting the inability to install `linkml` into their environment even though ostensibly it shouldn't be a problem (if, that is, our dependency spec wasn't so strict). Because of that history I'm just trying be very intentional about this kind of change.

For this specific issue I'd just point out that:

> if we know that <0.2 doesn't work for 3.12 then we should spec that here

We don't know that, not at all. When I said that versions of `prefixmaps` declare support for Python 3.12 starting in `0.2.0` that was just based on a cursory glance of the `classifiers` field in their `pyproject.toml` in recent version tags. And my point was just that requiring `prefixmaps` >=0.2.**2** on Python 3.12 was _definitely_ too restrictive.

> IMO it's good practice to exclude known bad dep combos

That's totally fair! But I think there is no _known_ bad combo here. I might have miscommunicated and caused some confusion there. If so, my bad.

Reply:
oh gotcha, i assumed we knew there was some problem with 3.12 just skimming the previous discussion on this PR. ya if we don't know there's an incompatibility, seems good to me!